File: linked_list_cycle.cpp
              //QUESTION: 141. Linked List Cycle


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
  */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast!= nullptr && fast -> next!= nullptr){
            slow = slow ->next;
            fast = fast ->next ->next;
            if(slow == fast){
                return true;
            }
        

        }
         return false;
    }
};

// my logic:
// class Solution {
// public:
//     bool hasCycle(ListNode *head) {

//         unordered_set<ListNode*> vis;

//         ListNode* t = head;

//         while(t) {

//             if(vis.find(t) != vis.end()) {
//                 return true;
//             }

//             vis.insert(t);
//             t = t->next;
//         }

//         return false;
//     }
// };

File: Dsa(6th-13th)june/DAY 7/Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode s = head;
        ListNode f = head;

        if (s == null || s.next == null || f.next.next == null) {
            return false;
        }

        s = s.next;
        f = f.next.next;

        while (f != s) {
            if (f == null || f.next == null) {
                return false;
            }

            s = s.next;
            f = f.next.next;
        }

        return true;
    }
}
